### PR TITLE
NPC frugal threshold gamerule.

### PR DESCRIPTION
**Pull request by 1010todd**
**Feature:** This PR implements the feature request detailed and discussed in issue none. Just found it while skimming the code and thought it could be a gamerule.

## Feature Details
Add "universal frugal threshold" as gamerule. I added "universal" because I imagine this could be expanded later where NPCs personality is more dynamic and less hardcoded and allow adjustment of frugal threshold per fleet/npc similar to what we currently do with `confusion` with the universal threshold as fallback if the threshold is not given. Another point I can see it being adjusted is player's escorts where they could adjust it to their preference (or even per-ship).

## UI Screenshots
N/A

## Usage Examples
```
gamerule
	"universal frugal threshold" .8
```
Where value ranges from 0. to 1. as percentage (so 0. = 0%, 1. = 100% and .8 in the example means 80% of total health)

## Testing Done
VSC doesn't say anything.

### Automated Tests Added
N/A

## Performance Impact
idk, hopefully little to none.

### CODE REVIEW

**Comment by Zitchas** [comment]
What does it do? Is that .8 in the example the amount of health it needs to be reduced to? The percentage of ammo it will use in an engagement before it meets its criteria to bring everything to bear?

**Comment by warp-core** [comment]
> What does it do? Is that .8 in the example the amount of health it needs to be reduced to? The percentage of ammo it will use in an engagement before it meets its criteria to bring everything to bear?

```
	# Ships with frugal personality will only use ammo-using weapon after their health drops below this percent threshold.
	# DEFAULT: .75 (75 percent)
	# ALLOWABLE VALUES: any double between 0. and 1.
```

@1010todd `double` isn't necessarily meaningful to someone looking to edit game data. Maybe `real number` or just `value`.

File: source/AI.cpp [hunk 1]
```diff
@@ -3231,9 +3231,9 @@ void AI::AutoFire(const Ship &ship, FireCommand &command, bool secondary) const
 	bool beFrugal = (ship.IsYours() && !escortsUseAmmo);
 	if(person.IsFrugal() || (ship.IsYours() && escortsAreFrugal && escortsUseAmmo))
 	{
-		// The frugal personality is only active when ships have more than 75% of their total health,
-		// and are not outgunned.
-		beFrugal = (ship.Health() > .75);
+		// The frugal personality is only active when ships have more than certain percent of their total health,
```
Comment (warp-core) [inline source/AI.cpp:3234]:
```suggestion
		// The frugal personality is only active when ships have more than a certain fraction of their total health,
```

File: source/Gamerules.cpp [hunk 2]
```diff
@@ -44,6 +44,8 @@ void Gamerules::Load(const DataNode &node)
 			noPersonSpawnWeight = max<int>(0, child.Value(1));
 		else if(key == "npc max mining time")
 			npcMaxMiningTime = max<int>(0, child.Value(1));
+		else if(key == "universal frugal threshold")
+			npcMaxMiningTime = min<double>(1,max<double>(0, child.Value(1)));
```
Comment (warp-core) [inline source/Gamerules.cpp:48]:
```suggestion
			universalFrugalThreshold = min<double>(1, max<double>(0, child.Value(1)));
```

File: data/gamerules.txt [hunk 3]
```diff
@@ -44,3 +44,8 @@ gamerules
 	# DEFAULT: 3600 frames (one minute)
 	# ALLOWABLE VALUES: any integer >= 0
 	"npc max mining time" 3600
+	
+	# Ships with "frugal" personality will only use weapons that consume ammo or fuel while their health is at or below this fraction.
```
Comment (warp-core) [inline data/gamerules.txt:48]:
```suggestion
	# Ships with the "frugal" personality will only use weapons that consume ammo or fuel while their health is at or below this fraction.
```

File: source/Gamerules.cpp [hunk 4]
```diff
@@ -44,6 +44,8 @@ void Gamerules::Load(const DataNode &node)
 			noPersonSpawnWeight = max<int>(0, child.Value(1));
 		else if(key == "npc max mining time")
 			npcMaxMiningTime = max<int>(0, child.Value(1));
+		else if(key == "universal frugal threshold")
+			universalFrugalThreshold = min<double>(1, max<double>(0, child.Value(1)));
```
Comment (warp-core) [inline source/Gamerules.cpp:48]:
```suggestion
			universalFrugalThreshold = min<double>(1., max<double>(0., child.Value(1)));
```

**Pull request merged**